File: src-map/tokenizer.cpp
#include "stemmer.cpp"
#include "trie.cpp"

std::vector<std::string> tokenize(const std::string& text) {
	std::vector<std::string> tokens;
	std::string temp;
	tokens.reserve(128);

	for (char ch : text) {
		if (std::isalpha((unsigned char)ch) || std::isdigit((unsigned char)ch)) {
			temp.push_back(std::tolower((unsigned char)ch));
		}
		else {
			if (!temp.empty()) {
				tokens.push_back(std::move(temp));
				temp.clear();
				temp.reserve(16);
			}
		}
	}
	if (!temp.empty())
		tokens.push_back(std::move(temp));
	return tokens;
}

std::vector<std::string> stem_tokens(const std::vector<std::string>& tokens) {
	std::vector<std::string> out;
	out.reserve(tokens.size());
	for (auto const& t : tokens) {
		std::string s = porter::get_stem(t);
		if (!s.empty()) out.push_back(std::move(s));
	}
	return out;
}

std::vector<std::string> get_tokens(const std::string& text) {
	auto toks = tokenize(text);
	return stem_tokens(toks);
}

File: src-java/Tokenizer.java
import java.util.*;

public class Tokenizer {
    public static List<String> tokenize(String text) {
        List<String> tokens = new ArrayList<>();
        StringBuilder temp = new StringBuilder();
        for (char ch : text.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                temp.append(Character.toLowerCase(ch));
            } else {
                if (temp.length() > 0) {
                    tokens.add(temp.toString());
                    temp.setLength(0);
                }
            }
        }
        if (temp.length() > 0) tokens.add(temp.toString());
        return tokens;
    }

    public static List<String> stemTokens(List<String> tokens) {
        List<String> out = new ArrayList<>();
        for (String t : tokens) {
            String s = Stemmer.getStem(t);
            if (!s.isEmpty()) out.add(s);
        }
        return out;
    }

    public static List<String> getTokens(String text) {
        List<String> toks = tokenize(text);
        return stemTokens(toks);
    }
}
